vsphere/internal/virtualdevice/virtual_machine_device_subresource.go

```diff
@@ -0,0 +1,745 @@
+	devices := l.Select(dsf)
+	if len(devices) != 1 {
+		return nil, fmt.Errorf("invalid device result - %d results returned (expected 1): controller key %q, disk number: %d", len(devices), ckey, du)
+	}
+	device := devices[0]
+	log.Printf("[DEBUG] FindVirtualDevice: Device found: %s", l.Name(device))
+	return device, nil
+}
+
+// FindVirtualDevice will attempt to find an address by its device key if it is
+// > 0, otherwise it will attempt to locate it by its device address.
+func (r *Subresource) FindVirtualDevice(l object.VirtualDeviceList) (types.BaseVirtualDevice, error) {
+	if key := r.Get("key").(int); key > 0 {
+		log.Printf("[DEBUG] FindVirtualDevice: Looking for device with key %d", key)
+		if dev := l.FindByKey(int32(key)); dev != nil {
+			log.Printf("[DEBUG] FindVirtualDevice: Device found: %s", l.Name(dev))
+			return dev, nil
+		}
+		return nil, fmt.Errorf("could not find device with key %d", key)
+	}
+	return r.FindVirtualDeviceByAddr(l)
+}
+
+// String prints out the device sub-resource's information including the ID at
+// time of instantiation, the short name of the disk, and the current device
+// key and address.
+func (r *Subresource) String() string {
+	devaddr := r.Get("device_address").(string)
+	if devaddr == "" {
+		devaddr = "<new device>"
+	}
+	return fmt.Sprintf("%s (key %d at %s)", r.Addr(), r.Get("key").(int), devaddr)
+}
+
+// swapSCSIDevice swaps out the supplied controller for a new one of the
+// supplied controller type. Any connected devices are re-connected at the same
+// device units on the new device. A list of changes is returned.
+func swapSCSIDevice(l object.VirtualDeviceList, device types.BaseVirtualSCSIController, ct string) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	log.Printf("[DEBUG] swapSCSIDevice: Swapping SCSI device for one of controller type %s: %s", ct, l.Name(device.(types.BaseVirtualDevice)))
+	var spec []types.BaseVirtualDeviceConfigSpec
+	bvd := device.(types.BaseVirtualDevice)
+	cspec, err := object.VirtualDeviceList{bvd}.ConfigSpec(types.VirtualDeviceConfigSpecOperationRemove)
+	if err != nil {
+		return nil, err
+	}
+	spec = append(spec, cspec...)
+
+	nsd, err := l.CreateSCSIController(ct)
+	if err != nil {
+		return nil, err
+	}
+	nsd.(types.BaseVirtualSCSIController).GetVirtualSCSIController().BusNumber = device.GetVirtualSCSIController().BusNumber
+	cspec, err = object.VirtualDeviceList{nsd}.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+	if err != nil {
+		return nil, err
+	}
+	spec = append(spec, cspec...)
+	ockey := device.GetVirtualSCSIController().Key
+	nckey := nsd.GetVirtualDevice().Key
+	for _, vd := range l {
+		if vd.GetVirtualDevice().ControllerKey == ockey {
+			vd.GetVirtualDevice().ControllerKey = nckey
+			cspec, err := object.VirtualDeviceList{vd}.ConfigSpec(types.VirtualDeviceConfigSpecOperationEdit)
+			if err != nil {
+				return nil, err
+			}
+			// Clear the file operation
+			cspec[0].GetVirtualDeviceConfigSpec().FileOperation = ""
```

Comment:
The implementation of [`ConfigSpec`](https://github.com/vmware/govmomi/blob/master/object/virtual_device_list.go#L891) is such that the amount of operations returned is stable and set to the number of devices you send in. I'm assuming it's stable, but I've added the assertion just in case.